### ISSUE TITLE
How can use onnx model for inference with AMD GPU in Windows?

### ISSUE THREAD

**Issue by francismelon** [microsoft/onnxruntime#19061]
### Describe the issue

My computer is Windows system, but only amd gpu, I want to use onnxruntime deployment, there are kind people can give me an example of inference, thank you very much!!

### To reproduce

- This is CUDA Example, but i wanna an AMD example.
```
void envInit()
{
    std::vector<const char*> inputNodeNames;
    std::vector<const char*> outputNodeNames;
    std::vector<int64_t> inputNodeDims;
    std::vector<int64_t> outputNodeDims;
    std::wstring onnxPath = "model.onnx";
    Ort::Session session;
    Ort::SessionOptions sessionOptions;
    Ort::MemoryInfo memoryInfo;
    Ort::Env env = Ort::Env{ORT_LOGGING_LEVEL_ERROR, "RunDetation"};

    OrtSessionOptionsAppendExecutionProvider_CUDA(sessionOptions, 0);
    sessionOptions.SetInterOpNumThreads(1);
    sessionOptions.SetGraphOptimizationLevel(GraphOptimizationLevel::ORT_ENABLE_ALL);
    session = Ort::Session(env, onnxPath.c_str(), sessionOptions);
    memoryInfo = Ort::MemoryInfo::CreateCpu(OrtDeviceAllocator, OrtMemTypeDefault);
    Ort::Allocator allocator(session, memoryInfo);
    inputNodeNames.push_back(session.GetInputName(0, allocator));
    outputNodeNames.push_back(session.GetOutputName(0, allocator));
    inputNodeDims = session.GetInputTypeInfo(0).GetTensorTypeAndShapeInfo().GetShape();
    outputNodeDims = session.GetOutputTypeInfo(0).GetTensorTypeAndShapeInfo().GetShape();

    // begin inference
    // .....
}
```

### Urgency

Its medium level urgent.

### Platform

Windows

### OS Version

10.0.22621.2861

### ONNX Runtime Installation

Released Package

### ONNX Runtime Version or Commit ID

1.12.1

### ONNX Runtime API

C++

### Architecture

X64

### Execution Provider

DirectML

**Comment by tianleiwu** [comment]
For 6500m GPU, you can try DirectML. 
Change the line
```
OrtSessionOptionsAppendExecutionProvider_CUDA(sessionOptions, 0);
```
to
```
OrtSessionOptionsAppendExecutionProvider_DML(session_options, 0)
```
See https://onnxruntime.ai/docs/execution-providers/DirectML-ExecutionProvider.html for more information

**Comment by francismelon** [comment]
Thank you for your help！！！I try to use DirectML, but I had other problems when i follow this steps of the Page.
- Download the onnxruntime1.12.1 package from this page and included in my project. 
![图片](https://github.com/microsoft/onnxruntime/assets/58884735/c773d3b1-1cea-4ab1-9d5d-bd0bc9876b31)
- Updated Windows 10 SDK version from 10.0.17763.0 to 10.0.18362.0 , i use vs2017 to compile it.
- Modified the code for DirectML Provider.
```
    std::vector<const char*> inputNodeNames;
    std::vector<const char*> outputNodeNames;
    std::vector<int64_t> inputNodeDims;
    std::vector<int64_t> outputNodeDims;
    std::wstring onnxPath = "model.onnx";
    Ort::Session session;
    Ort::SessionOptions sessionOptions;
    Ort::MemoryInfo memoryInfo;
    Ort::Env env = Ort::Env{ORT_LOGGING_LEVEL_ERROR, "RunDetation"};

    // Provider set DML
    OrtSessionOptionsAppendExecutionProvider_DML(sessionOptions, 0);
    sessionOptions.SetInterOpNumThreads(1);
    sessionOptions.SetGraphOptimizationLevel(GraphOptimizationLevel::ORT_ENABLE_ALL);
    
    // Configurations that must be modified to use DML
    sessionOptions.DisableMemPattern();
    sessionOptions.SetExecutionMode(ExecutionMode::ORT_SEQUENTIAL);

    session = Ort::Session(env, onnxPath.c_str(), sessionOptions);  // When I run the program, initializing the session fails on this line
    memoryInfo = Ort::MemoryInfo::CreateCpu(OrtDeviceAllocator, OrtMemTypeDefault);
    Ort::Allocator allocator(session, memoryInfo);
    inputNodeNames.push_back(session.GetInputName(0, allocator));
    outputNodeNames.push_back(session.GetOutputName(0, allocator));
    inputNodeDims = session.GetInputTypeInfo(0).GetTensorTypeAndShapeInfo().GetShape();
    outputNodeDims = session.GetOutputTypeInfo(0).GetTensorTypeAndShapeInfo().GetShape();
```
- Console print: Exception during initialization
`2024-01-10 15:47:42.4524398 [E:onnxruntime:, inference_session.cc:1530 onnxruntime::InferenceSession::Initialize::<lambda_300271ae079a93faf30b4ef2a83d2ebc>::operator ()] Exception during initialization:`

Can I get some advice from you?Anyway, thanks for your help!

> For 6500m GPU, you can try DirectML (It seems that your GPU is not supported by ROCm, otherwise ROCm provider might also work).
> 
> Change the line
> 
> ```
> OrtSessionOptionsAppendExecutionProvider_CUDA(sessionOptions, 0);
> ```
> 
> to
> 
> ```
> OrtSessionOptionsAppendExecutionProvider_DML(session_options, 0)
> ```
> 
> See https://onnxruntime.ai/docs/execution-providers/DirectML-ExecutionProvider.html for more information
> 
> 对于6500 m GPU，您可以尝试DirectML（ROCm似乎不支持您的GPU，否则ROCm提供程序也可能工作）。
> 
> 更改线
> 
> ```
> System. out. println（0）;
> ```
> 
> 到
> 
> ```
> OrtSessionOptiondExecutionProvider_DML（session_options，0）
> ```
> 
> 有关详细信息，请参阅[https：//onnxruntime.ai/docs/execution-providers/DirectML-ExecutionProvider.html](https://onnxruntime.ai/docs/execution-providers/DirectML-ExecutionProvider.html)

**Comment by tianleiwu** [comment]
@francismelon, What's the full message of the exception? 

You can try latest one: https://github.com/microsoft/onnxruntime/releases/tag/v1.16.3. It was built with Visual Studio 2022 14.38.33130, and Windows 10 SDK 10.0.22621.0.

**Comment by francismelon** [comment]
@tianleiwu, Thank you for your help :hand: !The version 1.12.1 I used only had the following exception message,which seems to be the whole message :
 ```
2024-01-10 15:47:42.4524398 [E:onnxruntime:, inference_session.cc:1530 onnxruntime::InferenceSession::Initialize::<lambda_300271ae079a93faf30b4ef2a83d2ebc>::operator ()] Exception during initialization:
``` 
---
I also feel very strange why is it empty, :joy: ! I think there is a relationship between the os version, the DirectML.dll version and the onnxruntime.dll version, and the libraries that depend on each other are different, so it doesn't create a `session` object. If I update the Windows 10 SDK and install VS2022, it will have a slight impact on my project environment.

---
But the code emample is worked in onnxruntime 1.11.1, DirectML.dll 1.8.0 and os 10.0.22621.2861. It is very cool!!!! :smiley: 
I also want to use in Windows 10 (10.0.19045.3803),but the program is crashed.Then I replaced the onnxruntime version and it didn't work until 1.3.0 and the performance was much slower than 1.11.1.Thank you for your advice!
